### PR TITLE
Rework Variant Selection Exceptions

### PR DESCRIPTION
**Pull request by tresat**
First followup to #27858. 

- Greatly consolidate the variant selection exception class hierarchy.
  - Now there are only 3 concrete types representing a failure selecting a component, a failure selecting an artifact variant, and a failure requesting a configuration by name.
  - Variant selection exceptions are now immutable - all resolutions and failure data is provided at construction.
- The `ResolutionFailure` implementation type now describes the particular failure.  This is stored as data in the thrown exception.  
  - The `ResolutionFailureHandlerIntegrationTest` has been reworked to expect the proper failure type in addition to the exception type, which will vary less.
  - The failure type is logged at the info level when a failure occurs.
- Remove all CC-incompatible types from failure data so that exceptions containing these types can be serialized.
  - Remove schema from all failure data, pass (optionally) schema to `ResolutionFailureDescriber`s as needed.
- `ResolutionFailureDescriber` reworked to take a single generic argument (the type of failure).

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

File: platforms/software/dependency-management/src/main/java/org/gradle/internal/component/resolution/failure/describer/InvalidMultipleVariantsFailureDescriber.java [hunk 1]
```diff
@@ -30,14 +30,14 @@
...
+public abstract class InvalidMultipleVariantsFailureDescriber extends AbstractResolutionFailureDescriber<IncompatibleMultipleNodeSelectionFailure> {
     private static final String INCOMPATIBLE_VARIANTS_PREFIX = "Incompatible variant errors are explained in more detail at ";
     private static final String INCOMPATIBLE_VARIANTS_SECTION = "sub:variant-incompatible";
 
     @Override
-    public IncompatibleArtifactVariantsException describeFailure(IncompatibleMultipleNodeSelectionFailure failure) {
+    public VariantSelectionException describeFailure(IncompatibleMultipleNodeSelectionFailure failure) {
         String msg = buildIncompatibleArtifactVariantsFailureMsg(failure);
-        IncompatibleArtifactVariantsException result = new IncompatibleArtifactVariantsException(msg);
+        VariantSelectionException result = new VariantSelectionException(msg);
```
Comment (jvandort) [inline platforms/software/dependency-management/src/main/java/org/gradle/internal/component/resolution/failure/describer/InvalidMultipleVariantsFailureDescriber.java:40]: Shouldn't this be an `ArtifactVariantSelectionException`?

File: platforms/software/dependency-management/src/main/java/org/gradle/internal/component/resolution/failure/type/AbstractConfigurationSelectionFailure.java [hunk 2]
```diff
@@ -16,27 +16,25 @@
...
 
     @Override
     public String getRequestedName() {
-        return requestedName;
+        return requestedConfigurationName;
     }
 
-    public ComponentIdentifier getRequestedComponent() {
-        return requestedComponent;
+    public String getRequestedComponentName() {
```
Comment (jvandort) [inline platforms/software/dependency-management/src/main/java/org/gradle/internal/component/resolution/failure/type/AbstractConfigurationSelectionFailure.java:37]: I would prefer we call this `getRequestedComponentDisplayName` or `getComponentDisplayName`. I think its important to differentiate an identifiable name vs some string used for reporting/logging

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/internal/component/ResolutionFailureHandlerIntegrationTest.groovy [hunk 3]
```diff
@@ -397,20 +392,20 @@ class ResolutionFailureHandlerIntegrationTest extends AbstractIntegrationSpec {
...
 
-    private final Demonstration incompatibleArtifactVariants = new Demonstration("Incompatible artifact variants", IncompatibleArtifactVariantsException.class, this.&setupIncompatibleArtifactVariantsFailureForProject)
-    private final Demonstration noMatchingArtifactVariants = new Demonstration("No matching artifact variants", NoMatchingArtifactVariantsException.class, this.&setupNoMatchingArtifactVariantsFailureForProject)
-    private final Demonstration ambiguousArtifactTransforms = new Demonstration("Ambiguous artifact transforms", AmbiguousArtifactTransformException.class, this.&setupAmbiguousArtifactTransformFailureForProject)
-    private final Demonstration ambiguousArtifactVariants = new Demonstration("Ambiguous artifact variants", AmbiguousArtifactVariantsException.class, this.&setupAmbiguousArtifactVariantsFailureForProject)
+    private final Demonstration incompatibleArtifactVariants = new Demonstration("Incompatible artifact variants", VariantSelectionException.class, this.&setupIncompatibleArtifactVariantsFailureForProject)
```
Comment (jvandort) [inline platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/internal/component/ResolutionFailureHandlerIntegrationTest.groovy:405]: This seems wrong, the previous failure was an artifact exception but now we've changed to a graph exception

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Review by big-guy — APPROVED** [review]

**Comment by tresat** [comment]
@bot-gradle test and merge

**Comment by tresat** [comment]
@bot-gradle test APT

**Comment by tresat** [comment]
@bot-gradle test APT

**Comment by tresat** [comment]
@bot-gradle test APT

**Comment by tresat** [comment]
@bot-gradle test APT

**Comment by tresat** [comment]
@bot-gradle test and merge

**Pull request merged**